pkg/kube/pod.go

```diff
@@ -114,16 +119,71 @@ func GetPodLogs(ctx context.Context, cli kubernetes.Interface, namespace, name s
 
 // WaitForPodReady waits for a pod to exit the pending state
 func WaitForPodReady(ctx context.Context, cli kubernetes.Interface, namespace, name string) error {
-	err := poll.Wait(ctx, func(ctx context.Context) (bool, error) {
+	timeoutCtx, waitCancel := context.WithTimeout(ctx, podReadyWaitTimeout)
+	defer waitCancel()
+	err := poll.Wait(timeoutCtx, func(ctx context.Context) (bool, error) {
 		p, err := cli.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
 		if err != nil {
 			return false, err
 		}
+
+		// check if nodes are up and available
+		n := strings.Split(p.Spec.NodeName, "/")
+		if n[0] != "" {
+			node, err := cli.CoreV1().Nodes().Get(n[0], metav1.GetOptions{})
+			if err != nil {
+				return false, errors.Wrapf(err, "Failed to get node %s", n[0])
+			}
+			if !IsNodeReady(node) || !IsNodeSchedulable(node) {
+				return false, errors.Errorf("Node %s is currently not ready/schedulable", n[0])
+			}
+		}
+
+		// check for memory or resource issues
+		if p.Status.Phase == v1.PodPending {
+			if p.Status.Reason == "OutOfmemory" || p.Status.Reason == "OutOfcpu" {
+				return false, errors.Errorf("Pod stuck in pending state, reason: %s", p.Status.Reason)
+			}
+		}
+
+		// check if pvc and pv are up and ready to mount
+		if err := checkPVCAndPVStatus(p, cli, namespace); err != nil {
+			return false, err
+		}
+
 		return p.Status.Phase != v1.PodPending && p.Status.Phase != "", nil
 	})
 	return errors.Wrapf(err, "Pod did not transition into running state. Namespace:%s, Name:%s", namespace, name)
 }
 
+func checkPVCAndPVStatus(p *v1.Pod, cli kubernetes.Interface, namespace string) error {
```

Comment:
please describe the intent of this function a bit - if the PVC is not ready, is that an error condition and we should return right away? Or does that mean we should wait for the timeout?  As-is this function mixes approaches. if `PersistentVolumeClaim` is nil for any volume, we return no error, but other empty conditions are treated as errors.

Reply:
So here we are checking if the PVC is stuck in pending state or has status =  `ClaimLost` .
If the PVC is stuck in pending state then we check the status of the PV associate with that PVC. If the PV is lost or has status failed then we return an error since the pvc will not transition into running state and hence the pods will also not transition into running state. otherwise we wait for the timeout.
if the status is `claimlost` then we return an error.